File: src/main/java/com/goldv/context/Context.java
package com.goldv.context;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

/**
 * Created by goldv on 27/04/2015.
 */
public class Context {

  private final Map<String, Object> scope;

  private final Stack<ChildContext> childContexts = new Stack<>();

  public Context(){
    this.scope = new HashMap<>();
  }

  public Context(Map<String, Object> scope){
    this.scope = scope;
  }

  public ChildContext deleteChild(){
    return childContexts.pop();
  }

  public void add(String field, Object value){
    scope.put(field, value);
  }

  public Map<String, Object> getScope(){ return scope; }

  public ChildContext createChild(){
    ChildContext cc = new ChildContext(this);
    childContexts.push(cc);
    return cc;
  }

  public Double getAsNumber(String field){
    Object value = getValue(field);
    if(value != null){
      if(value instanceof Integer) return ((Integer) value).doubleValue();
      else if(value instanceof Long) return ((Long) value).doubleValue();
      else if(value instanceof Double) return ((Double) value);
    }

    return null;
  }

  public String getAsString(String field){
    Object value = getValue(field);
    if(value != null){
      return value.toString();
    }

    return null;
  }

  public Collection<Object> getAsCollection(String field){
    Object value = getValue(field);
    if(value != null && value instanceof Collection){
      return (Collection)value;
    }

    return null;
  }

  public Boolean getAsBoolean(String field){
    Object value = getValue(field);
    if(value != null && value instanceof Boolean){
      return (Boolean)value;
    }

    return null;
  }

  private Object getValue(String field){
    // first check the child scopes
    Object childValue = getValueFromChildContext(field);
    if(childValue != null) return childValue;
    else return getValue(field, scope);
  }

  private Object getValueFromChildContext(String field){

    for( ChildContext child : childContexts){
      Object value = getValue(field, child.getScope());
      if(value != null) return value;
    }

    return null;
  }

  private Object getValue(String field, Map<String, Object> searchScope){

    int delimIdx = field.indexOf(".");
    if(delimIdx < 0){
      return searchScope.get(field);
    } else {
      String nextField = field.substring(0, delimIdx);

      Object next = searchScope.get(nextField);
      if(next != null && next instanceof Map){
        return getValue(field.substring(delimIdx + 1), ((Map) next));
      } else {
        return null;
      }
    }
  }
}

File: src/main/scala/com/goldv/context/Context.scala
package com.goldv.context



/**
 * Created by vince on 01/05/15.
 */
class Context(var context: Map[String, Any] = Map.empty[String, Any]) {

  var childContexts: List[Context] = Nil

  def createChild = {
    val ctx = Context()
    childContexts = ctx :: childContexts
    ctx
  }

  def deleteChild = {
    childContexts = childContexts.dropRight(1)
  }

  def put(field: String, value: Any) = context = context.updated(field, value)

  def getDouble(field: String) = getValue(field).flatMap{
    case n : Int => Some( n.doubleValue() )
    case n : Double => Some( n.doubleValue() )
    case n : Long => Some( n.doubleValue() )
    case n : Short => Some( n.doubleValue() )
    case n : Byte => Some( n.doubleValue() )
    case _ => None
  }

  def getBoolean(field: String) = getValue(field).flatMap{
    case b: Boolean => Some(b)
    case _ => None
  }

  def getString(field: String) = getValue(field).map(_.toString)

  def getList(field: String) = getValue(field).flatMap{
    case l: Seq[_] => Some(l.toList)
    case _ => None
  }

  def getValue(field: String): Option[Any] = getValueFromChildContext(field) match{
    case Some(childValue) => Some(childValue)
    case None => getValueFromContext(field, context)
  }

  def getValueFromChildContext(field: String): Option[Any] = childContexts.find(ctx => ctx.getValueFromContext(field, ctx.context).isDefined )
    .flatMap( ctx => ctx.getValueFromContext(field, ctx.context))

  def getValueFromContext(field: String, searchScope: Map[String, Any]): Option[Any] = {

    val delimIdx = field.indexOf(".")
    if(delimIdx < 0){
      searchScope.get(field)
    } else {
      val nextField = field.substring(0, delimIdx);

      searchScope.get(nextField).flatMap{
        case m:Map[_,_] => getValueFromContext(field.substring(delimIdx + 1), m.asInstanceOf[Map[String, Any]])
        case _ => None
      }
    }
  }

}

object Context{
  def apply() = new Context()
}
